### PR TITLE
Meta-analysis with MMUPHin and Multi-omics Integration with Cooperative Learning

### PR DESCRIPTION
**Pull request by YihanLiu4023**
New PR for 2 new chapters: meta-analysis with MMUPHin package and multi-omics integration with cooperative learning.

### CODE REVIEW

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
Hello!

First of all thank you for this PR. Looks very nice!!  Sorry for taking this long to respond. Unfortunately, I haven't had time to review this in detail. However with quick check, I noticed  some things that you might want to consider. I will get back to you when I have checked this more thoroughly.

-Tuomas

File: inst/pages/IntegratedLearner.qmd [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+```{r message=FALSE, warning=FALSE}
+
+#####################
+# Load example data #
+#####################
+
+library(tidyverse)
+library(curatedMetagenomicData)
+library(curatedMetagenomicAnalyses)
```
Comment (TuomasBorman) [inline inst/pages/IntegratedLearner.qmd:27]: Hello!

I started to review this, but ended up to issue. From where this "curatedMetagenomicAnalyses" package comes from? It seems that it is not in CRAN or Biconductor. Can it be avoided somehow?

File: inst/pages/IntegratedLearner.qmd [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+# Stack all models #
+####################
+
+# Run user-specified meta learner
+SL_fit_stacked<-SuperLearner::SuperLearner(Y = Y, 
+                                           X = combo, 
+                                           cvControl = cvControl,    
+                                           verbose = TRUE,
+                                           SL.library = 'SL.glmnet',
+                                           family = binomial())
```
[the remark this answers is not in the document]
Comment (YihanLiu4023) [reply inst/pages/IntegratedLearner.qmd:298]: I've already fixed all the problems you mentioned above, please see the changed files. I've also add the SuperLearner to the DESCRIPTION file, can you see the changes successfully? Please let me know if I did not make the changes in the right place. Thank you!

[the remark this answers is not in the document]
Comment (TuomasBorman) [reply inst/pages/IntegratedLearner.qmd:298]: Thanks, I've will check it now in more detail

[the remark this answers is not in the document]
Comment (TuomasBorman) [reply inst/pages/IntegratedLearner.qmd:298]:
```
Error in `makeSEforCondition()`:
! could not find function "makeSEforCondition"
```

I cannot read the data

[the remark this answers is not in the document]
Comment (YihanLiu4023) [reply inst/pages/IntegratedLearner.qmd]: I've re-uploaded the data for IntegratedLearner and MMUPHin. Please check again, thank you!

[the remark this answers is not in the document]
Comment (YihanLiu4023) [reply inst/pages/IntegratedLearner.qmd]: Besides, I've also uploaded another chapter called "MSEA.qmd" for the PR, thank you!

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
1. All the packages that are used, should be in DESCRIPTION file
2. Some chunk options were incorrect

I think with these changes this is good to go

**Comment by antagomir** [comment]
Please resolve the review discussions when they have been handled. This way we can keep track on possible issues here that are still pending. When no discussion points are open it is a good sign that this is getting ready.

File: inst/pages/IntegratedLearner.qmd [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+multiview analysis. Proc Natl Acad Sci U S A. 2022;119(38):e2202113119. 
+doi:10.1073/pnas.2202113119.
+2.  van der Laan MJ, Polley EC, Hubbard AE. Super learner. Stat Appl Genet Mol 
+Biol. 2007;6:Article25. doi:10.2202/1544-6115.1309.
+3.  Mallick H, Porwal A, Saha S, Basak P, Svetnik V, Paul E. An integrated 
+Bayesian framework for multi-omics prediction and classification. Stat Med. 
+Published online December 26, 2023. doi:10.1002/sim.9953.
+4.  Lloyd-Price J, Arze C, Ananthakrishnan AN, et al. Multi-omics of the gut 
+microbial ecosystem in inflammatory bowel diseases. Nature. 2
+019;569(7758):655-662. doi:10.1038/s41586-019-1237-9.
```
[the remark this answers is not in the document]
Comment (antagomir) [reply inst/pages/IntegratedLearner.qmd:506]: Yes, please move the references to the bib file

File: inst/pages/IntegratedLearner.qmd [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+all(rownames(feature_table) == rownames(feature_metadata)) # TRUE
+all(colnames(feature_table) == rownames(sample_metadata)) # TRUE
+
+```
+
+Further data pre-processing is necessary to handle near-zero-variance features in this dataset. Ultimately, 483 features are retained, consisting of 360 pathways and 123 species. A combination of variance and prevalence filtering is applied to the feature table, while related metadata is cleaned to ensure the retention of matching samples.
+
+```{r create-feature-table}
+
+library(caret)
```
[the remark this answers is not in the document]
Comment (antagomir) [reply inst/pages/IntegratedLearner.qmd:129]: caret is very common, might be unavoidable in the end so might be easier to include it

File: inst/pages/IntegratedLearner.qmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (antagomir) [reply inst/pages/IntegratedLearner.qmd]: Close this discussion if these issues have been resolved. @YihanLiu4023 I think you can close if you can confirm that you fixed them.

**Comment by antagomir** [comment]
Only very small final changes needed, it seems to me.

@YihanLiu4023 - let us know if you need any support from our side in finalizing these. This is a good opportunity to get in sync with the general development guidelines.

File: inst/pages/IntegratedLearner.qmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (TuomasBorman) [reply inst/pages/IntegratedLearner.qmd:129]: That's fine, we can check which dependencies can be dropped in the future

**Comment by antagomir** [comment]
Hi! I see see that you're working on this @YihanLiu4023 - just confirm when ready, so we know to proceed checking.

**Comment by YihanLiu4023** [comment]
> Hi! I see see that you're working on this @YihanLiu4023 - just confirm when ready, so we know to proceed checking.

Sure! Thank you for the reminder, I have already uploaded all the updates. Please feel free to proceed the checking.

**Review by TuomasBorman — APPROVED** [review]
I think this is now good to go. Some changes related to build still needed but I can do them. 

These chapters are very nice! Thanks!

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+and apply the PERMANOVA test.
+
+```{r batch-effect-evaluation-pathway-abundance}
+
+D_before <- vegdist(t(data_abd))
+D_after <- vegdist(t(data_abd_adj))
+
+set.seed(1)
+fit_adonis_before <- adonis(D_before ~ study_name, data = data_meta)
+fit_adonis_after <- adonis(D_after ~ study_name, data = data_meta)
```
Comment (antagomir) [inline inst/pages/MMUPHin_meta_analysis.qmd:386]: See above.

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+                 paste('Axis 2 (', percent_var_before[2], '%)', sep = ''))
+
+before_phrase <- paste('Before (PERMANOVA R2 = ', R2_before, '%)', sep = '')
+colnames(ord_before) <- c('PC1', 'PC2')
+ord_before$Study <- data_meta$study_name
+
+# After
+R2_after <- round(fit_adonis_after$aov.tab[1, 5]*100, 1)
+pcoa_after <- cmdscale(D_after, eig = TRUE)
+ord_after <- as.data.frame(pcoa_after$points)
```
Comment (antagomir) [inline inst/pages/MMUPHin_meta_analysis.qmd:418]: same

**Review by antagomir — CHANGES_REQUESTED** [review]

**Comment by antagomir** [comment]
Ok thanks. Now with the more detailed look I noticed one major issue, which is that the scripts are primarily using base R & vegan, rather than the TreeSummarizedExperiment framework. This is somewhat problematic since the entire OMA gitbook is built on the idea of supporting the standardization of methods and shift towards the TreeSummarizedExperiment ecosystem of tools. 

Now I think we have two options:

1) Update the current PR to use SE framework, before merging

2) Merge now, and then update afterwards, ideally by the original contributor @YihanLiu4023 and/or with support from the broader contributor base. This latter option has the benefit that we can proceed with other things that are pending on merging this PR but this also has a risk that these necessary updates towards SE ecosystem will be delayed.

@YihanLiu4023 @TuomasBorman your comments will be appreciated

**Comment by antagomir** [comment]
Ok I can see that this was merged already so I guess we are going with the option 2.

@YihanLiu4023 would you be able to make some of the suggested updates (above pending comments) in a new PR?

**Comment by antagomir** [comment]
If the idea is to update this to IntegratedLearning and that will support SE/MAE, then we can deal with these in the next round?

**Comment by antagomir** [comment]
@YihanLiu4023 could you summarize the planned updates and their planned schedule?

**Comment by antagomir** [comment]
I just noticed that these new chapters are not rendered as part of the OMA gitbook https://microbiome.github.io/OMA/

I think the reason is that the qmd file names have not been listed here: 
inst/assets/_book.yml  

@YihanLiu4023 did you check that you can render and view the book locally, before opening the PR?

**Comment by TuomasBorman** [comment]
This issue is already handled with PR https://github.com/microbiome/OMA/pull/433/files

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+fit_adjust_batch <- adjust_batch(feature_abd = data_abd,
+                                 batch = "study_name",
+                                 covariates = "resvar",
+                                 data = data_meta,
+                                 control = list(verbose = FALSE))
+
+data_abd_adj <- fit_adjust_batch$feature_abd_adj
+
+```
+![](https://github.com/microbiome/OMA/tree/master/inst/pages/images/adjust_batch_diagnostic_pathway_abundance)
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:374]: @YihanLiu4023 It seems like this can not generated by the script which is used internally within the package. Maybe we can exclude it?

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+                                 batch = "study_name",
+                                 covariates = "resvar",
+                                 data = data_meta,
+                                 control = list(verbose = FALSE))
+
+data_abd_adj <- fit_adjust_batch$feature_abd_adj
+
+```
+![](https://github.com/microbiome/OMA/blob/master/inst/pages/images/adjust_batch_diagnostic_relative_abundance)
+
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:171]: Thanks for the comment. We have now removed the plot as it's not easy to generate the plot using the script which is used internally within the package.

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+Based on the results, we can see that the study differences can explain a total
+of 11.922% ($R^2$) of the variability in microbial abundance profiles before
+study effect adjustment, whereas after adjustment this was reduced to 2.806%
+($R^2$).
+
+Let's visualize the results of the PERMANOVA test. 
+```{r visualization}
+
+library(ggplot2)
+library(cowplot)
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:215]: Sounds great. Yes, this can be done.

File: inst/pages/MSEA.qmd [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+  returnSamples("pathway_abundance", rownames = "short")
+
+##########################
+# Create sample metadata #
+##########################
+
+sample_metadata <- 
+  colData(se_pathway) %>% 
+  as.data.frame() %>% filter(visit_number == 1) %>% 
+  select("age", "disease", "antibiotics_current_use")
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MSEA.qmd:362]: Previously we deleted the `::` operator but it seems unavoidable. We will modify it accordingly.

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+
+```{r load-pkg-data}
+
+######################
+# Load melanoma data #
+######################
+
+library(dplyr)
+
+se_relative <- sampleMetadata |>
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:100]: Good point. Thanks for the comment!

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+
+# Convert the "study_name" to factor variable
+data_meta$study_name <- as.factor(data_meta$study_name)
+
+###########################
+# Create Species Features #
+###########################
+# Transpose the abundance matrix and change the value of abundance data to proportion unit
+data_abd <- assay(se_relative)
+data_abd <- data_abd / 100
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:151]: I would suggest removing this line where we are dividing by 100 because the results will be unchanged and modify the chunk title accordingly.

File: inst/pages/MMUPHin_meta_analysis.qmd [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+
+# Match the individuals in the data_abd
+data_abd <- data_abd[, colnames(data_abd) %in% rownames(data_meta)]
```
[the remark this answers is not in the document]
Comment (himelmallick) [reply inst/pages/MMUPHin_meta_analysis.qmd:154]: It might be a bit of work to achieve that given that we are in a multi-study setup and we are working on the metadata first to extract the individual studies and the associated changes that are independent of the assay data. Any workaround is appreciated.

**Comment by himelmallick** [comment]
I agree that going with Option 2 and updating later. We will make sure to use the SE/TSE framework in the `IntegratedLearner` R package which can be added back to the chapters instead of the base R functions and the same thing applies to the other packages. Thanks all for your comments. @YihanLiu4023 will work on the immediately addressable comments from above.

**Comment by antagomir** [comment]
great to hear, we needed to merge but let's continue and take it forward together; don't hesitate to ask if anything is unclear as the ecosystem and documentation are shaping up and we can improve where necessary
